### DISCUSSION TITLE
[react] Types for 19

### DISCUSSION THREAD

**Discussion by eps1lon** [DefinitelyTyped/DefinitelyTyped#64451] [Issues with a @types package]
React 19 is not released. This issue is meant to collect type-related breaking changes we want to ship with the next major:

1. [[react]: Infer state type for useReducer](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64412)
1. default `P` of `ReactElement` to `unknown`
   I missed to do this in React 18. It'll also help mitigate https://github.com/microsoft/TypeScript/issues/53178. See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64840
1. (codemod: `refobject-defaults`) refs are mutable by default. `RefObject` will not include `null` by default: [[react] Make all refs mutable by default](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64896)
1. (codemod: `useref-required-initial`) Require initial value for `useRef`: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64920
1. (codemod: `scoped-jsx`) Remove deprecated global `JSX` namespace: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64464
1. Remove deprecated `React.SFCFactory`
1. (codemod: `deprecated-react-text`) Remove deprecated `React.ReactText`
1. (codemod: `deprecated-react-child`) Remove deprecated `React.ReactChild`
1. (codemod: `deprecated-react-node-array`) Remove deprecated `React.ReactNodeArray`
1. Remove deprecated `ReactFragment`
1. (codemod: `deprecated-void-function-component`) Remove deprecated `React.VFC`
1. (codemod: `deprecated-void-function-component`) Remove deprecated `React.VoidFunctionComponent`
1. Remove deprecated `React.ReactChildren`
1. Remove deprecated `React.ComponentSpec` (use `createReactClass.ComponentSpec` instead)
1. Remove deprecated `React.Mixin` (use `createReactClass.Mixin` instead)
1. Remove deprecated `React.ClassicElement<Props>` (use `createReactClass.ClassicElement<Props, InstanceType<T>>`
1. Remove deprecated `React.ClassicComponentClass` (use `createReactClass.ClassicComponentClass`)
1. Remove deprecated `React.ClassicComponent` (use `createClass.ClassicComponent`)
1. (codemod: `deprecated-prop-types-types`) Remove deprecated `React.Requireable` (use `propTypes.Requireable`)
1. (codemod: `deprecated-prop-types-types`) Remove deprecated `React.Validator` (use `propTypes.Validator`)
1. (codemod: `deprecated-prop-types-types`) Remove deprecated `React.ValidationMap` (use `propTypes.ValidationMap`)
1. (codemod: `deprecated-prop-types-types`) Remove deprecated `React.WeakValidationMap` (use `propTypes.WeakValidationMap`)

<details>
<summary>Type changes to reflect runtime changes</summary>

- [x] `enableFormActions` and `enableAsyncActions`
- [x] rendering `bigint`, `Promise` and ~React Context~ (see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69347/ for why no context yet)
- [x] new DOM Transition event handlers
- [x] `enableNewBooleanProps`
- [x] `requestFormReset`
- [x] `onCaughtError`, `onUncaughtError` 
- [x] Remove `createFactory`
- [x] Remove `defaultProps` on everything but class components
- [x] ~Remove~ Stop typing `propTypes`
- [x] Remove `render`
- [x] Remove `findDOMNode`
- [x] Remove `renderToNodeStream`
- [x] Remove `hydrate`
- [x] Remove `unmountComponentAtNode`
- [x] Remove `unstable_renderSubtreeIntoContainer`
- [x] Remove everything from `test-utils` but `act`
- [x] Remove `ErrorInfo#digest`
- [x] `enableFloat`
- [x] Switch `<Context>` to mean `<Context.Provider>`
- [x] `enableRefAsProps`
- [x] `ref` cleanup functions
- [x] `React.act`
- [x] Remove `react-test-renderer/shallow`
</details>

codemods: https://github.com/eps1lon/types-react-codemod/issues/314
Work-in-progress PR (changes to to 3rd party libraries using deprecated changes will be cherry-picked separately to keep the diff small): Pending

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Added "default `P` of `ReactElement` to `unknown`"

**Comment by Semigradsky** [comment]
This one? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64099

**Reply by eps1lon** [reply]
Not sure about this one since this apparently didn't catch any real-word issues. But it will require props changes where people accept mutable props even though they only use it as readable. 

So I sympathize but TypeScript is all about making pragmatic decisions considering type-safety.

**Comment by eps1lon** [comment]
Added https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64896

**Comment by eps1lon** [comment]
Added scoped instead of global namespace: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64464

**Comment by peterblazejewicz** [comment]
fyi: pinned for improved transparency

**Comment by eps1lon** [comment]
Added `Classic*` types deprecated in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/68091

**Comment by ghost** [comment]
Deprecated `ReactComponentElement` type in #68457

**Comment by eps1lon** [comment]
Added deprecation of types related to `propTypes` in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69002

**Comment by vincent-thomas** [comment]
Will forwardRef and the default prop 'ref' in components be addressed here?

**Reply by eps1lon** [reply]
> the default prop 'ref' in components be addressed here?

Yep.

What should we adress about `forwardRef`?

**Comment by mayank99** [comment]
Since React 19 adds full support for custom elements, it would be nice to not raise errors on elements that don't have explicitly defined types. They should just be treated like any regular/unknown `HTMLElement`. It's valid HTML after all, and the custom elements may be defined by someone else outside of React.

The desired behavior would be similar to custom attributes that have a dash in them and don't raise TS errors.

![](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/9084735/7d24ed57-419d-4396-8f18-11d8f3159de3)

**Reply by eps1lon** [reply]
> It's valid HTML after all, and the custom elements may be defined by someone else outside of React.

You should custom elements via module augmentation: [Playground Link](https://www.typescriptlang.org/play/?jsx=4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQLABQ9AJlrgDbKVwgSMCurWcpTwwycAN7040uADtkILKjB5BAKQDKADQlSZ+4LJhYoBVXACSRqIdTBcAUQGKj6SXX2eZZEAE8AtLi8qDAk-ljOWEZkAFy6Hl6J0gBuyKy8WHEhNrIA5npJ0gC+BZ4lCcX05fQwvmCCAIJwALyYVDAAdJpaHVYwOXaOka709AD0Y3AAAjCo4QAe9fjhUMRQcAQQEHDA6LIQ8MhwqazAjHBgxGD0ADyMwMkbW3BjAHwA3OOTM3NYiywwFZrOAAWV2djyFyucAABicMjDbn5AsFQiBwsN4G9PnQbsigiEwhEsC54PCsM0AEQQWRYSkvD5AA). 

We'll continue to error on excess props e.g. `<div foo />` since there's no way to distinguish typos from excess props. If you want to add non-standard attributes, check out [Playground Link](https://www.typescriptlang.org/play/?jsx=4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQLABQ9AJlrgDbKVwgSMCurWcpTwwycAN7040uADtkILKjB5BAKQDKADQlSZ+4LJhYoBVXACSRqIdTBcAUQGKj6SXX2eZZEAE8AtLi8qDAk-ljOWEZkAFy6Hl6J0gBuyKy8WHEhNrIA5npJ0gC+BZ4lCcX05fQwvmCCAIJwALyYVDAAdJpaHVYwOXaOka709AD0Y3AAAjCo4QAe9fjhUMRQcAQQEHDA6LIQ8MhwqazAjHBgxGD0ADyMwMkbW3BjAHwA3OOTM3NYiywwFZrOAAWV2djyFyucAABicMjDbn5AsFQiBwsN4G9PnQbsigiEwhEsC54PCsM0AEQQWRYSkvD5AA)

**Reply by mayank99** [reply]
I know we can use module augmentation, but I'm saying it feels wrong to raise TS errors on custom elements that don't do this.

The whole point of custom elements is that they're framework-agnostic. Most custom elements don't (and shouldn't) come with React-specific type declarations. Since these custom elements work totally fine in React at runtime, they shouldn't result in red squiggles inside the IDE; it makes it look like React doesn't support custom elements. Just treat them like regular HTML elements (e.g. `<div>`) that don't accept any special props.

To be clear about attributes, I'm not suggesting changing unknown props. I'm just pointing out that attribute names with dashes don't raise TS errors, even though they don't have explicitly defined types.

**Reply by mayank99** [reply]
Without built-in TS support for these platform features, I have to do this in every new React project:

```ts
// fix-react.d.ts
declare module 'react' {
  namespace JSX {
    interface IntrinsicElements {
      [key: `${string}-${string}`]: React.HTMLAttributes<HTMLElement>;
    }
  }
  interface CSSProperties {
    [key: `--${string}`]: string | number;
  }
}
```

[Playground Link](https://www.typescriptlang.org/play/?jsx=4#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwCwAUIwCZa4A2ylcIEzAruyzlKeGGTgBvRnFlwAdshBZUYPMIBSAZQAaUmXMPB5MLFALq4ASRNRjqYLgCiQ5SfTSGh73IDaAaywATwAuOAADABJJVBg7eQBzAF8AWmjY+OTwgF0w7DEAOgAJABUAWQAZAEEYOOAAI35TVAAeUsqXLDcYAD56Lx84JINZYYHZY1NzSwBhLS0ABWIwMxhgFX1x7wDgsPCUtJi6xKScsIzjBLgAHwV+EHqzfsMxscYWkCCUrFcsEx6WgB6T7fX7-fofL4-Lp-eAg5C1KAAXgARAA3ZDsfhYFEA4FQsG9CEg3D8WIkUEwkxwWJBIRIySSOAog6k8kgFFhFHNGAooZJOCAvqMRiAwFwAACMFQ3wAHit8N8oMQoO8SWSYBTod04BisVhURB5DjBcKmAwxZLpXKFTAlSr3gQIBBTf1ReKpTKsPK2HazA6GC1mMA0TSYHSDYy4E6IFz6lw+UkBUKaHBGEA)

**Reply by zeorin** [reply]
I'd much rather be able to decide how to let TS know about how React can use my custom elements, because the example above might silence errors but at the same time it also makes it easy to forget to add type definitions that are precise, e.g.: 

```typescript
import 'types-react' // types: rc
import 'types-react-dom' // types: rc

type Color = NonNullable<React.CSSProperties["color"]>

class MyColorElement extends HTMLElement {
  inner = this.attachShadow({ mode: "open" })
  color?: Color

  static get observedAttributes() {
    return ["color"]
  }

  constructor() {
    super()
  }

  attributeChangedCallback(name: string, oldValue: unknown, newValue: unknown) {
    if (name === "color" && oldValue !== newValue) {
      const element = this.inner.querySelector("input")
      if (!element) return
      element.value = newValue as string
      this.color = newValue as string
    }
  }

  connectedCallback() {
    const color = this.getAttribute("color")

    this.inner.innerHTML = /* html */ `
      <label>
        <span>Colorize!</span>
        <input value="${color ?? ""}">
        <button>Apply</button>
      </label>
    `

    this.inner.querySelector("button")?.addEventListener("click", () => {
      const value = this.inner.querySelector("input")?.value

      if (!value) return

      this.color = value

      const payload = new CustomEvent("ColorChange", {
        detail: value,
        bubbles: true,
        composed: true,
      })

      this.dispatchEvent(payload)
    })
  }
}

window.customElements.define("my-color", MyColorElement)

export const MyColor = "my-color"

interface MyCustomEvent<D = unknown, C = unknown, T = unknown>
  extends CustomEvent<D> {
  currentTarget: C & EventTarget
  target: T & EventTarget
}

interface ColorChangeEvent<T = Element> extends MyCustomEvent<Color, T> {}

type ColorChangeEventHandler<T = Element> = {
  bivarianceHack(event: ColorChangeEvent<T>): void
}["bivarianceHack"]

interface MyColorHTMLAttributes<T> extends React.HTMLAttributes<T> {
  color?: Color | undefined
  onColorChange?: ColorChangeEventHandler<T> | undefined
  onColorChangeCapture?: ColorChangeEventHandler<T> | undefined
}

export interface MyColorJSXIntrinsicElement {
  "my-color": React.DetailedHTMLProps<
    MyColorHTMLAttributes<MyColorElement>,
    MyColorElement
  >
}

declare module "react" {
  namespace JSX {
    interface IntrinsicElements extends MyColorJSXIntrinsicElement {}
  }
}

const color: Color = "black";

<MyColor
  color={color}
  onColorChange={(event) => {
  	event.target
  	//    ^?
  	event.currentTarget
  	//    ^?
  	event.currentTarget.color
  	//    			    ^?
  	event.detail
  	//    ^?
  }}
  onClick={(event) => {
  	event.target
  	//    ^?
  	event.currentTarget
  	//    ^?
  	event.currentTarget.color
  	//                  ^?
  }}
/>;

type MyColorRef = React.ComponentRef<typeof MyColor>
//   ^?

<my-color
  color={color}
  onColorChange={(event) => {
  	event.target
  	//    ^?
  	event.currentTarget
  	//    ^?
  	event.currentTarget.color
  	//    			    ^?
  	event.detail
  	//    ^?
  }}
  onClick={(event) => {
  	event.target
  	//    ^?
  	event.currentTarget
  	//    ^?
  	event.currentTarget.color
  	//                  ^?
  }}
/>;

type My_ColorRef = React.ComponentRef<'my-color'>
//   ^?
```

[TypeScript Playground](https://www.typescriptlang.org/play/?target=99&jsx=4#code/JYWwDg9gTgLgBAchgTzAUwM4FopoIYDGMCcA9KXCuhgFxxQEBQoksiVmO+RWAJhCBLlKqTHQaNGHOAGEIAG2hwAvHAByEAHZqArvPl4ARvLQAeAErcYAOhkBlOwAUoEdLGCYA2gCICC6N4AugB8kgQGGBhwALLIcopQAKImIGia8GgAHjBpvFEAEgAq0QAyyWip6XAA3oxwcMCammhQKpQAFsAY1ngwMITtdu14-ADuABTVcCAQvGh03q5p3nAAvgCUdXB+CQD8dPHQkvUY-TDABHAA5mjwEIYYLQBuaLwAgn1QwIY6ORjj6xqW3quBgOigmjgPh2AUCW1Wx22WlOUB0RGgAKB9ROOjcAPhiN6MC+PxyMmGmhuvBkeH0hkIAGtxpo8Kk6CjGlcADRwBS8ABqtJ08zgOk0DM0EFGmh5zVGgvkwroYolUs0gNq2IaADM4MzWWgVMpVL5-FAVgAyC28+QCoWGgCExrgcoVwo1wOxfk0pzgaBSaXgqhgnW6jWaUGsAEdhVBkHZ-Wh0VBxt5GmBft5Nlr6sBdeMHYnKjBAaDwZpPfUi4HrE97W1XfW8FEOZTKx0utYYa1VI3FYbm3BW1dPQj6gitt7mkRXjS6YzMZqvcj4N22iHOzcYB9id9fmhU92s4j6huw00WtZwy0iqU2qQAFRwdowEDyOAPigAA3bpgMhn9UIc2xUwMDAPBNGCQ4vgALzQB1TFIMCIKA4C4FMdNfjgOt+2UbwABJqjXXZdjgbxvFWbxUOA0xSRgLRgjeMAwHkZBELohjf1If9AM9H9PTPK8L0jGMWnjRNk1TDjNCzXYel4XhEhedISi6HII0PeQLgZbweUxZRgixHNvV9HDhXXUMhIjaNY3EkxJLTTQMxgWTa3tE9sTzPUHTMtBS1ucsPNPSy11UXygqRH14HA5BFBGBs0FGWQdFOAQlMDVNoPJCCbl0ozgLmfpgHkOhfK5dt6h+QxjDEShUTQcq0KRcAIEeXg6GJYVGpzDYIsE3gunAmACHadL0nGGK4t4bNsV68dGAnUZGjGLsUvokBymLbo5m1RoD28EBkCwI8eViaDNsDTZGCyVhVxXGI4jNNoDqOo9JEaHIoG1QhDTOta0uUmBTAAETaFVJWlHkZDB8UIZlOBChh1VpVQrJ1LyZLUo2wGQcMpcCHBXB0kKPAoC3A44GtMaYBJsnbi2fo6ZgOhEapwHaa3Bb3vSFpvoIQ0sopG5qdMRHVAu9JDLR3Ioj+rGRegnlCjxidpEFnK0Gp-IIN4EwoFFtoJZgQzVCXQxgDrL4IP57WCCZNBAYOM1sspTWceV9ZSogYBeAWnxzct4BrbQW2dLhZgea+n6Hug28Sh3El9wwUWpeyGW4EsQgbDjhO9z+FOjO7fZZCegAfUVNB2vbffqLR1dd4v6+FwHtcrvWC-LsUq+aGveU0Ju0BpMAwVwRvnaFt3A1b3WWg7ivu9eLnrsyW6GkjvnfsehIACk7AADQASXSL4fQuI2jJe46zW8OhM6IaxgduPBiteOPnFcZPPTOs0c8+PPMFMN-BIRtgjdXqEA6ARstihAnHMcIpNDQzF4HoQ03hcBZxWEuFkqRkL8zgLvPe+U16fQ3nAI+u5T4ECNlEaWldZZb2gAQ8hJ8MBnwDFUaoY41hLxMndBITsEjPWMIybwABuSQgCGFQEnGaZQREzRcLruPDWcjxgO0uioPGWwACQ6j0jWEZpzeo2jhDYgAHq7B0XomwBMoBExpqTIxcATEUHMZY4x1jVp2MDBzW4XYzQ6NMcY7R2itQWKsYDawhVn7yECa4+o4TxyKP7lpO2qjrGAgMkZXRkTDH02MUEuAiTnGeNsfY3xMA4lhPcSUyJZSfGOL8d2KpTVWnFNWAiUgwRxFSFEDHM0lhdSqDvjYOQLVmjpEGaYDgEBdQQKgKEIJ4TGCmEOlfBIMiEhyO7Mkge6TAaZK0R43JjTKkFPiUUmpOSaz1OJqclpCSrmlMJg0pm-iNnnK1CE0JbiIk1micVB5lz4S7NSQyfZGislLmufovJZznGFOKTCmxLy7lMyBUi553i0VbneUcT5rSmrtM6d0yQ0hYgAH1oKDLaCM2wAhIATJgFMhAazuwIEWfEixQA)

Similarly, your CSS custom properties example also isn't super precise. See https://github.com/frenic/csstype#what-should-i-do-when-i-get-type-errors.

Now, that might be fine for you. But to make that the default experience for everyone is not the right call, IMO.

**Reply by mayank99** [reply]
I never suggested making my exact code the default experience for everyone. You should absolutely be able to define more precise types if you want. See [forked playground](https://www.typescriptlang.org/play/?target=99&jsx=4#code/JYWwDg9gTgLgBAchgTzAUwM4FopoIYDGMCcA9KXCuhgFxxQEBQoksiVmO+RWAJhCBLlKqTHQaNGvNAQA2eXHBAReAV1lo4AIlyEYWuAG9GcOADs8ITGEKaAUgGUAGkZOm4wMzDRQAZrbgASS8oTwxgAgBRDSsvDFd3dwBtAGs0ZDoAAwASQwwYULMAcwBfLFz8wtLMgF06ACVuGAA6ABE0GDxgDV4ACQAVAFkAGQAFKAgwDAAeRr1mgZGAQRgC4AAjVW8ZxeHotFiYAD4AGl39w6OAbjdErRBkLAIIWWgtBqa2jq6e3fHJmaDZAAYRe0F2KzWm220yBoNeUAuaC8pzhYMRMWRx1ucBKbjxeMYwn6AAtgPFyXBfJ5NIBeDcAsjuMaYPLAAdxJeG8ADcfHBeJy8FhfBAIABeLTrBQGUjXSTCcYycmaKieIoeeL5bqyOCQDDhdYaOAMpks54Itxm6Dig2EFJaNwQMzw6DAjnFNCiwwACjQPK8AEo4KKjgk4ABIX1Y5oEVRQXBefoKIodaPotxh4SJLPZrMAPQA-OnI15mtJOt105nTAX8XiZTciRQAAIwbBoAAe6B4PgmUDopMpmtk2p70HiEC2RsZzMelqgVJF4q0R2mpFN6NljbgLbbnZkMCwo77cAH8TwsgwEDgQ5HcbHcAn8GNM7ZHO5vOFYq00s3W6w-4AwCgOAkDQLA8CIMgoDJA4OBnXnUU4AAOUdJD1HkA00FmT5gQcBx-nQWBgEwJItDnLQaiOSQ5DwPU4DRBEkS8OAO28MxeHic5MWY4xTE8MxeUQmAyQwZpOU6AgSQcDl+FZL1DCUFQ0DoLRJmRAwSn9C10XzOh4MkUx8k5CI4GTeAIHWDAfB5XhIVCaFMC9QNePcXAYFjMw4FI8ianxAy4GeMxKlUIhoCc0NDNUQinL8txxPsrY0FdPB3V4YFz1kSUCBSL0LCsOhKlVE4H1kXgADVz1UZS4FUMwUjMCBWTMYqBNZCrZCquhavqxqzGcnFgF8OBcssTRRXG7RyLgAAyaaSvKyrNAAQgm1r2qq-rs0C-IWO4+AhJE5p+J8ZoAEcqqgZAHDQDRQqgL0tE8MAti0LTs0G4alpug4sUDNyPJxUxvsOZouUWoNzDQNrwdo681mKQHKEOucIbWmGNXhoocTxUxCVMQKBKINA0oyrKcs29xtvgFGDvJZozLsjZEoe8itJxYS6eOqAjrMASoF2CHSAAKjgEkYBAbUhYoTJEemDCbqonM4GmDAbDMI54OAAAvNAltXVWUsVnNpieycwY6j0tFyFH83zbQtBKZdEdMaZoRgR0jiWMAwFkZBVzdj3ZdIeXZCN0wZfZw6ubOi6ru+u6HoDsxXvzMTeF4SI-RgYZyTYnwWdkCI7WK8LgwiynHR282qohjnROj86fDj273fux6zGe-R-VT6u0H8xIPq9Jbe7+joAcRuvUwRCHe-7iugvgGxkFePBeFRqG4NUfIBEzrEHvg5L3S0YqXOzMsfjoXuTmduBNnWTDaEoKAquvpXnnACArN4OgChfxHNLnqYSevByQ2BgJJXeXgvRLxXrwN67gAG40YISVknhZLRi3u7EATFWyljQNSASD11wImPvREE6IcFs08N4PwAQ4SYJ3lnaYrQIbdQak1YqwJWF1XYc1E83CepNSNqxZEHFN7b2wUw1oIYXIxjjFiRMUAzJ6RmnASBMBFFmTcJ0JRHR+yqPUZojoyDJDUJ8P4AgmgD5umTOo6Y-QIY4JDCI9i8R6ESLsfBYq-QZGElgtYlKtis69BSrwDQUB7GOL2iGRCLl1jADBqEFKliQnZR9FnPS6JD5BKxPYo4-pL4QGALwZBpF4mJOAMktAqS7S+WYF4cxdDyEIghKsBKMIfEsXbGxMRcwiALCGMMRmDkZidNkTpTJ08AA+NV2L4JpCU0wjoAnul0nBLJNi0DqJCexcJeS4AzNqtIAhxMHROg2YEpKeAwDuVwGslZOSvA7LCT4fZhy5knJKSUIAA).

All I'm saying is that raising these type errors by default is undesirable and honestly a bad look for React. It's equivalent to if (hypothetically) all data attributes raised type errors by default, forcing everyone to type them precisely or add workarounds. I expect data attributes, custom elements, and custom CSS properties to all work out-of-the-box without errors, since these are all platform features specifically designed to be extended by authors.

**Reply by eps1lon** [reply]
>  You should absolutely be able to define more precise types if you want.

But your suggestion would default everybody to have loose types with regards to DOM attributes. 

>  It's equivalent to all data attributes raising type errors by default, forcing everyone to type them precisely or add workarounds.

`data-*` attributes do not raise type errors by default (e.g. `<div data-whatever="my-choice" />`. There's only one bug regarding them tracked in https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/69864 and waiting for a contributor.

**Reply by mayank99** [reply]
> `data-*` attributes do not raise type errors by default

Yes, I'm saying that custom elements raising type errors is equivalent to *if* data attributes also did the same.

Isn't it basically the same situation? By default, everyone has loose types (but no type errors) for data attributes.

**Reply by eps1lon** [reply]
> Isn't it basically the same situation? By default, everyone has loose types (but no type errors) for data attributes.

`data-*` attributes have the `string` type by (HTML) design. `data-*` attributes are intended for extending behavior. That doesn't mean you should also allow any attributes on every Element.

If we loosen the types, consumers cannot make them stricter. And even if they could, it's probably not the default most people want. 

On the flipside, the behavior you want can be implemented today:

```tsx
import * as React from 'react'

declare module 'react' {
    interface HTMLAttributes<T> {
        [anyProp: string]: unknown
    }

    namespace JSX {
        interface IntrinsicElements {
            [anyElement: string]: Record<keyof any, unknown>
        }
    }
}

<div foo />;
<this-doesnt-exist />;

// Probably meant `type` but is now permitted by types
<button typ="submit" />;
```

[Playground Link](https://www.typescriptlang.org/play/?jsx=4&ssl=5&ssc=35&pln=3&pc=25#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQLABQ9AJlrgDbKVwgSMCurWcpTwwycAN7040uMAB2MLFAJ5BACQAqAWQAyAQRgwowAEa9FqADwaAfBKkzHAbWRyAngAViYAFxxURvIA5gC6frxyANZyEADucg7SAL70iXByyCBYqGCqcABSAMoAGvZ0jo7yisp5AJIKxnKowLgAogJZCuiS5RV9Lu7tWJ0wfgGNoX7YuNCMlpFYbhAEKO4ANHAR0XFyNmmOKb3J9If0lozAAG5wBBAQcAD0NgDcZzAAFsCoALSMENkKb5YAAeX3gT1eDDoDwecC8EBMyBMrDcXBwCjgAAMYG4wFhMXAzPAvuk4nA8VAQMBDFhGITUTi8agzkSYBA5HBGQBeABEqF4JipMB5jxeQA)

**Comment by dylanhu7** [comment]
Is the `action` prop for the `form` element meant to be able to accept an `async` function (or more specifically a `(formData: FormData) => void | Promise<void>`)?

It seems to be the case according to the [current React docs](https://react.dev/reference/react-dom/components/form#props), but I couldn't find it reflected in the latest types. I'm not sure how or by whom the `DO_NOT_USE_OR_YOU_WILL_BE_FIRED_EXPERIMENTAL_FORM_ACTIONS` interface is meant to be used and if the framework (e.g. Next.js) is supposed to augment that.

**Reply by eps1lon** [reply]
> Is the action prop for the form element meant to be able to accept an async function (or more specifically a (formData: FormData) => void | Promise<void>)?

Yes, it just happened to work because of TypeScript quirks where functions with any return type are assignable to functions with the `void` return type. 

Added explicit support in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70197 to not rely on this quirk.

**Comment by trusktr** [comment]
What do we install for react 19 types? I see latest `@types/react` is 18.

**Reply by trusktr** [reply]
Ah nvm, I see here: https://react.dev/blog/2024/04/25/react-19-upgrade-guide#installing, we need to map the `types-react` package.

**Comment by acdvorak** [comment]
React `v19.0.0` was just released an hour ago: https://www.npmjs.com/package/react/v/19.0.0 😀

**Reply by eps1lon** [reply]
And so are the types: https://www.npmjs.com/package/@types/react/v/19.0.0

Landed in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69022

**Comment by GuichiZhao** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/72820
Is the bug fixed in React 19?
